consensus/src/chained_bft/event_processor.rs

```diff
@@ -709,6 +747,14 @@ impl<T: Payload> EventProcessor<T> {
                 error!("Error inserting qc {}: {:?}", qc, e);
                 return None;
             }
+
+            // Safety invariant: qc_grandparent <-- qc_parent <-- qc
+            // the preferred block round must be at least as large as qc_grandparent's round.
+            debug_checked_verify!(
```

Comment:
I'm not sure this check is correct: aren't we updating the preferred round only after `process_certificates()`? (should move this check after the `process_certificates()` invocation)

Reply:
The safety invariant here checks that the `preferred_block_round` is greater or equal to the `grandparent` of the `current proposal's qc` (If this condition is not satisfied, the replica should not vote for this proposal). `process_certificates()` will update this value to be the `parent` of the `current proposal's qc`.

Reply:
If it turns out that this is not correct, I would dearly love to see test cases that fail on this invariant.

Reply:
The safety invariant here checks that the `preferred_block_round` is greater than or equal to the `grandparent` of the `current proposal's qc` (If this condition is not true, then according to the voting rules, the replica should not add a vote). `process_certificates()` will update  `preferred_block_round` to the `parent` of the current proposal's qc.

Reply:
The `add_vote()` function is not about proposal processing and not about the voting decisions: it's about vote aggregation. The question is whether it's possible to aggregate enough votes in order to gather QC s.t. QC's grandparent round would be greater than your current preferred block round. I just don't understand the reason we have this verification here. What are we trying to check? :)

Reply:
Ideally, I  would like this check to be made right before sending a vote inside `process_proposed_block()`. However, this check requires access to `safety_rules` which is not accessible from `process_proposed_block()`. 
Since `add_vote` is also called from `process_remote_timeout_msg`, I added this check here. Do you have a better suggestion for placing this check?

Reply:
Ideally, I would like this check to be made right before sending a vote inside `process_proposed_block()`. However, this check requires access to `safety_rules` which is not accessible from `process_proposed_block()`.
Since `add_vote` is also called from `process_remote_timeout_msg`, I added this check here. Do you have a better suggestion for placing this check?

Reply:
I must confess I don't fully understand the rationale of this check. What invariant are we trying to enforce? One possible invariant could be that in the end of `process_certificates()` function we verify that preferred_block_round is == block_store.highest_qc.parent_round. But this is something else :)

Reply:
>One possible invariant could be that in the end of `process_certificates()` function we verify that preferred_block_round is == block_store.highest_qc.parent_round. But this is something else :)

Actually, this was the invariant I was trying to enforce. Thanks to my limited understanding of the data structures, it ended up in the wrong location and hence the confusion. After looking at your comments, I realized that `consensus_state` was accessible from `process_proposed_block()`. I have moved the invariant there. It should be straightforward to understand now.